packages/handler-express/src/createExpressSynthqlHandler.ts

```diff
@@ -52,38 +58,53 @@ async function executeSynthqlRequest<DB>(
     res: Response,
 ) {
     // First try to parse the request body as JSON
-    const { query, returnLastOnly } = await tryParseRequest(req);
+    const { body, headers } = await tryParseRequest(req);
 
     // We don't do this yet, but eventually we'll want to validate the request
     // const validatedQuery = await tryValidateSynthqlQuery(query);
 
     // Execute the query, but just to get the initial generator
-    const resultGenerator = await tryExecuteQuery<DB>(
-        queryEngine,
-        query,
-        returnLastOnly,
-    );
+    const resultGenerator = isRegisteredQueryRequest(body)
+        ? await tryExecuteRegisteredQuery<DB>(
+              queryEngine,
+              { queryId: body.queryId, params: body.params },
+              headers.returnLastOnly,
+          )
+        : isRegularQueryRequest(body)
+          ? await tryExecuteQuery<DB>(
+                queryEngine,
+                body.query,
+                headers.returnLastOnly,
+            )
+          : await tryExecuteQuery<DB>(
+                queryEngine,
+                body,
+                headers.returnLastOnly,
+            );
 
     // Now that we have the generator, we want to iterate over the items
     // and depending on `returnLastOnly`, we will write the status code
     // either before, or after iteration
-    await writeBody(res, query, resultGenerator, returnLastOnly);
+    await writeResponseBody(res, body, resultGenerator, headers.returnLastOnly);
 
+    // End response stream
     res.end();
 }
 
 async function tryParseRequest(req: Request) {
-    const body = req.body;
-    const returnLastOnly = req.headers['x-return-last-only'] === 'true';
-
     try {
-        const query = JSON.parse(body);
-
-        return { query, returnLastOnly };
+        return {
+            body:
+                typeof req.body === 'string' ? JSON.parse(req.body) : req.body,
+            headers: {
+                ...req.headers,
+                returnLastOnly: req.headers['x-return-last-only'] === 'true',
```

Comment:
Nothing to change here, but in a next release we should probably move x-return-last-only to the HTTP response.